crates/types/src/predicate.rs

```diff
@@ -24,3 +30,98 @@ pub struct Predicate {
     )]
     pub constraints: Vec<ConstraintBytecode>,
 }
+
+impl Predicate {
+    /// Maximum number of state read programs of a predicate.
+    pub const MAX_STATE_READS: usize = u8::MAX as usize;
+    /// Maximum size of state read programs of a predicate in bytes.
+    pub const MAX_STATE_READ_SIZE_BYTES: usize = 10_000;
+    /// Maximum number of constraint check programs of a predicate.
+    pub const MAX_CONSTRAINTS: usize = u8::MAX as usize;
+    /// Maximum size of constraint check programs of a predicate in bytes.
+    pub const MAX_CONSTRAINT_SIZE_BYTES: usize = 10_000;
+    /// Maximum size of a predicate in bytes.
+    pub const MAX_BYTES: usize = 1024 * 50;
+
+    /// Iterator over the programs in the predicate.
+    pub fn as_programs(&self) -> impl Iterator<Item = &[u8]> {
+        self.state_read
+            .iter()
+            .chain(self.constraints.iter())
+            .map(|x| x.as_slice())
+    }
+
+    /// An owning Iterator over the programs in the predicate.
+    pub fn into_programs(self) -> impl Iterator<Item = Vec<u8>> {
+        self.state_read.into_iter().chain(self.constraints)
+    }
+
+    /// Generate the encoding header for this predicate.
+    pub fn encoded_header(&self) -> Result<header::EncodedHeader, PredicateError> {
+        (self).try_into()
+    }
+
+    /// Encode the predicate into a bytes iterator.
+    pub fn encode(&self) -> Result<impl Iterator<Item = u8> + '_, PredicateError> {
+        let header = self.encoded_header()?;
+        Ok(header
+            .into_iter()
+            .chain(self.as_programs().flat_map(|x| x.iter().copied())))
+    }
```

Comment:
Maybe it's worth having an intermediary `encode_chunks`?

```rust
    pub fn encode_chunks(&self) -> Result<impl Iterator<Item = &[u8]> + '_, PredicateError> {
        let header = self.encoded_header()?;
        Ok(header
            .into_iter()
            .chain(self.as_programs()))
    }
```

It's probably technically more efficient to work with these slices directly than a fully flattened iterator, but rustc probably works optimisation magic in many cases anyways.

Reply:
This is possible using the `as_programs` method and calling the encode header function if you really need it but it's likely never needed. This function is specifically for when you need a flat vector of bytes. I benchmarked a bunch of different approaches and it really makes zero difference what you do even up to the max predicate size. Every time you run the bench you get a different winner. Basically it all just becomes calls to memcpy. If someone can create a bench that shows a meaningful difference I'll definitely be open to changing this but honestly I wasted a lot of time trying to optimise different approaches only to find all roads lead to memcpy

Reply:
Also note that the lifetimes don't work when you try to chain a local owned type with a reference to slices. I tried really hard to find a way around this problem but there's nothing good and it doesn't pay off in perf anyways